adapters/repos/db/search.go

```diff
@@ -267,23 +267,41 @@ func (db *DB) objectSearch(ctx context.Context, offset, limit int,
 	totalLimit := offset + limit
 	// TODO: Search in parallel, rather than sequentially or this will be
 	// painfully slow on large schemas
-	db.indexLock.RLock()
-	for _, index := range db.indices {
-		// TODO support all additional props
-		res, _, err := index.objectSearch(ctx, totalLimit,
-			filters, nil, sort, nil, additional, nil, tenant, 0)
-		if err != nil {
-			db.indexLock.RUnlock()
-			return nil, errors.Wrapf(err, "search index %s", index.ID())
-		}
+	// wrapped in func to unlock mutex within defer
+	if err := func() error {
+		db.indexLock.RLock()
+		defer db.indexLock.RUnlock()
 
-		found = append(found, res...)
-		if len(found) >= totalLimit {
-			// we are done
-			break
+		for _, index := range db.indices {
+			// TODO support all additional props
+			res, _, err := index.objectSearch(ctx, totalLimit,
+				filters, nil, sort, nil, additional, nil, tenant, 0)
+			if err != nil {
+				// TODO find better way to recognise particular errors
+				if errors.As(err, &objects.ErrInvalidUserInput{}) {
+					// validation failed (either MT class without tenantKey or non-MT class with tenantKey)
+					if strings.Contains(err.Error(), "has multi-tenancy enabled, but request was without tenant") ||
+						strings.Contains(err.Error(), "has multi-tenancy disabled, but request was with tenant") {
```

Comment:
make constants out of those errors?

Reply:
There is also `error.Is` which might make the code easier (but error handling in go suuuuuucks)

Reply:
I saved it for later. Definitely that deserves better impl. I consider that temporary solution